libchisel/src/checkexports.rs

```diff
@@ -0,0 +1,188 @@
+use super::ModuleValidator;
+
+use parity_wasm::elements::{
+    ExportEntry, ExportSection, External, FunctionSection, FunctionType, ImportSection, Internal,
+    Module, Type,
+};
+
+/// Enum representing a type of export and any extra data to check.
+pub enum ExportType<'a> {
+    Function(&'a str, FunctionType),
+    Global(&'a str),
+    Memory(&'a str),
+    Table(&'a str),
+}
+
+/// Trait over ExportType that lets a caller check if it is exported in a given module.
+trait IsExported {
+    fn is_exported(&self, module: &Module) -> bool;
+}
+
+/// Struct on which ModuleValidator is implemented.
+struct CheckExports<'a> {
+    entries: Vec<ExportType<'a>>,
+    allow_unlisted: bool,
+}
+
+impl<'a> CheckExports<'a> {
+    fn ewasm() -> Self {
+        CheckExports {
+            entries: vec![
+                ExportType::Function("main", FunctionType::default()),
+                ExportType::Memory("memory"),
+            ],
+            allow_unlisted: false,
+        }
+    }
+}
+
+impl<'a> ModuleValidator for CheckExports<'a> {
+    fn validate(self, module: &Module) -> Result<bool, String> {
+        // FIXME: This validating algorithm runs in O(n^2). Needs to be optimized
+        let required_exports_not_found = self
+            .entries
+            .iter()
+            .map(|x| x.is_exported(module))
+            .find(|x| *x == false)
+            .is_some();
+        // TODO: check if any extra exports exist
+
+        let module_export_count = if let Some(section) = module.export_section() {
+            section.entries().len()
+        } else {
+            0
+        };
+
+        let extra_exports_ok = self.allow_unlisted || (self.entries.len() == module_export_count);
+
+        Ok(!required_exports_not_found && extra_exports_ok)
```

Comment:
I think these could be simplified:
```
let required_exports_ok = !required_exports_not_found;

if (!required_exports_ok)
  return Ok(false);

if (self.entries.len() != module_export_count)
  Ok(self.allow_unlisted)
```

Reply:
Yeah I prefer the localized return but this also works

Reply:
Yeah but it is hard to follow what's going on. I'm not sure the `extra_exports_ok` line is correct.